pkg/k8s/client/restConfig_provider.go

```diff
@@ -215,3 +262,137 @@ func (r *restConfigManager) WrapRoundTripper(rt http.RoundTripper) http.RoundTri
+	if finfo, err := os.Stat(K8sAPIServerFilePath); errors.Is(err, os.ErrNotExist) {
+		_, err := os.Create(K8sAPIServerFilePath)
+		if err != nil {
+			return fmt.Errorf("unable to create '%s': %w", K8sAPIServerFilePath, err)
+		}
+	} else if finfo.Size() != 0 {
+		// Restore the service and endpoint addresses.
+		r.updateK8sAPIServerURL()
+	}
+
+	watcher, err := fswatcher.New([]string{K8sAPIServerFilePath})
+	if err != nil {
+		return fmt.Errorf("unable to add file watcher '%s': %w", K8sAPIServerFilePath, err)
+	}
+
+	r.jobs.Add(job.OneShot("kube-apiserver-state-file-watcher", func(ctx context.Context, health cell.Health) error {
+		health.OK("Starting kube-apiserver state file watcher")
+		return r.k8sAPIServerFileWatcher(ctx, watcher, health)
+	}))
+
+	return nil
+}
+
+func (r *restConfigManager) k8sAPIServerFileWatcher(ctx context.Context, watcher *fswatcher.Watcher, health cell.Health) error {
+	for {
+		select {
+		case <-ctx.Done():
+			health.Stopped("Context done")
+			watcher.Close()
+			return nil
+		case event := <-watcher.Events:
+			if !event.Op.Has(fsnotify.Write) {
+				continue
+			}
+			r.log.WithField("file", K8sAPIServerFilePath).Info("Processing write event ")
+			r.updateK8sAPIServerURL()
+		case err := <-watcher.Errors:
+			health.Degraded(fmt.Sprintf("Failed to load  %q", K8sAPIServerFilePath), err)
+			r.log.WithError(err).Error("Unexpected error while watching kube-apiserver state file")
+			return err
+		}
+
+	}
+}
+
+func (r *restConfigManager) updateK8sAPIServerURL() {
+	f, err := os.Open(K8sAPIServerFilePath)
+	if err != nil {
+		r.log.WithError(err).WithField(logfields.Path, K8sAPIServerFilePath).Error("unable " +
+			"to open file, agent may not be able to fail over to an active kube-apiserver")
+	}
+	defer f.Close()
+
+	var mapping K8sServiceEndpointMapping
+	if err = json.NewDecoder(f).Decode(&mapping); err != nil {
+		r.log.WithError(err).WithFields(logrus.Fields{
+			logfields.Path: K8sAPIServerFilePath,
+			"entry":        mapping,
+		}).Error("failed to " +
+			"decode file entry, agent may not be able to fail over to an active kube-apiserver")
+	}
+	if err = r.checkConnToService(mapping.Service); err == nil {
+		r.log.WithField("host", mapping.Service).Info("Updated kubeapi server url host")
+		// Set in tests
+		mapping.Service = strings.TrimPrefix(mapping.Service, "http://")
+		r.rt.Lock()
+		defer r.rt.Unlock()
+		r.rt.apiServerURL.Host = mapping.Service
+		r.Lock()
+		defer r.Unlock()
+		r.restConfig.Host = mapping.Service
+		updatedServerURLs := make([]*url.URL, 0)
+		for _, endpoint := range mapping.Endpoints {
+			endpoint = fmt.Sprintf("https://%s", endpoint)
+			serverURL, err := url.Parse(endpoint)
+			if err != nil {
+				r.log.WithError(err).Errorf("Failed to parse endpoint %s, skipping", endpoint)
+				continue
+			}
+			updatedServerURLs = append(updatedServerURLs, serverURL)
+		}
+		if len(updatedServerURLs) != 0 {
+			r.apiServerURLs = updatedServerURLs
+		}
+	}
+}
```

Comment:
nit: Why is this writting this way? Why do not a early return with `err != nil` and have the body of this if be the normal code path?